base/strings/io.jl

```diff
@@ -354,7 +354,8 @@ function join(io::IO, iterator, delim="")
 end
 
 function _join_preserve_annotations(iterator, args...)
-    if isconcretetype(eltype(iterator)) && !_isannotated(eltype(iterator)) && !any(_isannotated, args)
+    et = @default_eltype(iterator)
+    if isconcretetype(et) && !_isannotated(et) && !any(_isannotated, args)
```

Comment:
Should `et` be used instead of the `eltype` call in the other branch as well?

Reply:
Yes, that's probably sensible. I pushed the change